rfcs/git-sdk-requirements/RFC.md

```diff
@@ -0,0 +1,122 @@
+# RFC git sdk requirements
+
+## Background
+
+This RFC evaluates two architectures on building a git backend that serves current and foreseen needs of inlang.
+
+The inlang editor currently uses the git JS implementation. The JS implementation lacks git features that are required to build out the editor. The lack of git features triggered the exploration of switching the underlying git implementation, see [https://github.com/inlang/inlang/issues/278](https://github.com/inlang/inlang/issues/278).
+
+### Glossary
+
+- **isomorphic git** -The JavaScript git implementation currently used in the inlang editor.
+- **libgit2** - The git implementation that can be compiled to WebAssembly.
+- **git-sdk** - Client side git implementation that takes over cloning a repository.
+- **git-server** - Server that host git repositories. Think of GitHub.
+
+## Goals
+
+- Focus on iteration speed.
+  - We deal with high uncertainty. The faster we can iterate, the better we can react to changing requirements.
+- Keep external usage of the backend in mind.
+  - How easy can developers use the git backend? We intend to offer version control as a backend to external developers.
+
+## Non-goals
+
+- Architect the perfect system based on currently known requirements.
+  - The requirements will change. We can’t architect the perfect system now.
+
+## Requirements
+
+### Must run in the browser/on the client [High Confidence]
+
+Solving git limitations client-side, in the git-sdk, simplifies application architectures and is large part of inlang’s “the next git” thesis.
+
+Problems we are running into are identical on both the server and client. For example, reducing the bandwidth/time to clone a repository: Moving git-sdk logic to the server only shifts the problem to the server, but doesn’t solve it. The client would need to wait for the server to finish cloning of the repositories. Instead, the git-sdk can be adjusted to support lazy loading of files and thereby automatically solve long cloning times client- and server side.
+
+### Lazy loading of files and git history [High Confidence]
+
+Oftentimes only a subset of files in git repositories are required. Only cloning files and their history that is needed substantially reduces network and storage requirements. A lazy loading solution is likely coupled with the filesystem implementation.
+
+```jsx
+/**
+ * An example lazy loading implementation API.
+ */
+
+// does not clone the entire repository.
+// only metadata that enable other git commands to run
+await clone("https://github.com/inlang/inlang", fs)
+
+// lazy fetches the file and git commit history of README.md
+const commitHistory = await history("/README.md", fs)
+
+// lazy fetches xyz file
+const readme = await fs.readFile("/xyz.md")
+```
```

Comment:
This is a nice API and idea. 

As I read this it becomes more clear to me that the answer to the RFC should be whatever codebase we can modify easier. Isomorphic Git is clear winner here compared to libgit2 written in C. 

As another option I thought of compiling [go-git](https://github.com/go-git/go-git) to WASM via [TinyGo](https://tinygo.org). Go code is much easier to extend than C.  

However I am confident now that Isomoprhic Git should allow us to build this kind of lazy loading feature and more without disregarding web performance and velocity of adding new features.

Reply:
> answer to the RFC should be whatever codebase we can modify easier. 

I also realized that. Iteration speed should be the focus. Having a heavy toolchain that includes compiling C or GO to WASM will slow down iteration speeds. 

> However I am confident now that Isomoprhic Git should allow us to build this kind of lazy loading feature and more without disregarding web performance and velocity of adding new features. 

I am looking forward to your completion of the RFC! :)